hinted handoff: Use host IDs instead of IPs in the module

dawmd opened the pull request:
This pull request introduces host ID in the Hinted Handoff module. Nodes are now identified by their host IDs instead of their IPs. The conversion occurs on the boundary between the module and `storage_proxy.hh`, but aside from that, IPs have been erased.

The changes take into considerations that there might still be old hints, still identified by IPs, on disk – at start-up, we map them to host IDs if it's possible so that they're not lost.

Refs #6403 
Fixes #12278

bhalevy commented on db/hints/manager.cc:
```diff
@@ -446,6 +448,62 @@ locator::host_id manager::convert_ep_to_host_id(endpoint_id ep) const {
+
+            if (!ep_or_hid.has_host_id()) {
+                // This either finds the corresponding host ID or throws an exception.
+                ep_or_hid.resolve(*tmptr.get());
```
why is `.get()` needed?

bhalevy commented on db/hints/manager.cc:
```diff
@@ -446,6 +448,62 @@ locator::host_id manager::convert_ep_to_host_id(endpoint_id ep) const {
+                // See the System Calls Manual and this post:
+                //  https://stackoverflow.com/questions/7054844/is-rename-atomic
+                // for more information.
+                return io_check(rename_file, ep_path.c_str(), hid_path.c_str());
```
I prefer `.native()` rather than `.c_str()` since we're passing it to seastar, not to the kernel (so we don't strictly need a "C" interface)

bhalevy commented on db/hints/manager.cc:
```diff
@@ -446,6 +448,62 @@ locator::host_id manager::convert_ep_to_host_id(endpoint_id ep) const {
+                return io_check(rename_file, ep_path.c_str(), hid_path.c_str());
+            }
+
+            return make_ready_future<>();
```
you can use a coroutine lambda and call `co_await io_check` above and `co_await lister::rmdir` below.

bhalevy commented on db/hints/manager.cc:
```diff
@@ -446,6 +448,62 @@ locator::host_id manager::convert_ep_to_host_id(endpoint_id ep) const {
+            // We only care about directories named after IPs during an upgrade.
+            // In normal circumstances, this whole lambda should be equivalent to a no-op,
+            // i.e. it should have no side effects.
+            return lister::rmdir(datadir / de.name);
```
need to print an error in this case.

gusev-p commented on service/storage_proxy.cc:
```diff
@@ -3037,7 +3059,8 @@ storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::tok
 
-    if (cannot_hint(all, type)) {
+    // TODO: Use small_vector here instead to avoid dynamic allocations.
+    if (cannot_hint(to_host_id(all, get_token_metadata_ptr()), type)) {
```
we should use `erm->get_token_metadata()` since this is what we used to construct natural_endpoints and pending endpoints

gusev-p commented on db/hints/manager.cc:
```diff
@@ -223,76 +234,83 @@ future<> manager::wait_for_sync_point(abort_source& as, const sync_point::shard_
 
-bool manager::too_many_in_flight_hints_for(endpoint_id ep) const noexcept {
+bool manager::too_many_in_flight_hints_for(locator::host_id hid) const noexcept {
+    const auto ep = convert_host_id_to_ep(hid);
```
We can't just use the current `token_metadata` from `storage_proxy`, it may have changed and doesn't contain host_id anymore.  We can use `is_me` version with `host_id`. With gossiper we could probably check that there is at least one endpoint among `get_nodes_with_host_id` which is `<= _max_hint_window_us`.

gusev-p commented on service/storage_proxy.cc:
```diff
@@ -4033,14 +4058,20 @@ void storage_proxy::send_to_live_endpoints(storage_proxy::response_id_type respo
         db::hints::manager& hints_manager = hints_manager_for(type);
-        return boost::count_if(targets, [&mh, tr_state = std::move(tr_state), &hints_manager] (gms::inet_address target) mutable -> bool {
-            return mh->store_hint(hints_manager, target, tr_state);
+        const auto tmptr = get_token_metadata_ptr();
```
Same as in other places, we should use captured token_metadata from erm. In this place we can just remove this change if we switch `per_destination_mutation::_mutations` to `host_id`-s.

gusev-p replied:
OK

gusev-p commented on db/hints/manager.hh:
```diff
@@ -308,21 +310,33 @@ private:
     }
+
+private:
+    endpoint_id convert_host_id_to_ep(locator::host_id hid) const;
```
The function with this signature can't be implemented. The `token_metadata` may change and this function would fail on left `host_id`-s.

dawmd commented:
Version 2 changelog:
1. I made the changes "more compact", i.e. there are fewer but bigger commits.
2. I use proper `locator::effective_replication_map_ptr`s for mappings IP–host ID.
3. I applied most of Petr's suggestions. I didn't migrate `per_destination_mutation` to using `locator::host_id`; instead I pass `locator::effective_replication_map_ptr` to `mutation_holder` in storage proxy. The reason is that the API is mostly used for performing regular mutations that still rely on IPs. Doing so would require far more changes and most likely slow down the whole software, so I decided not to go with it.
4. We sync the parent directory after renaming each folder with hints.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5768/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5768/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7137/)
#### Failed Tests (4/23715):
- [test_hintedhandoff_removenode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5768/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_removenode) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_removenode)
- [test_hintedhandoff_dont_revive](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5768/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_dont_revive) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_dont_revive)
- [test_hintedhandoff_rebalance](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5768/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_rebalance) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_rebalance)
- [test_hintedhandoff_basic_check](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5768/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_basic_check) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_basic_check)
#### Build Details:
- Duration: 1 hr 57 min
- Builder: spider1.cloudius-systems.com

dawmd commented:
The failures in dtests come from the fact that they still assume hint directories are identified using IPs.

dawmd commented on db/hints/manager.cc:
```diff
@@ -316,98 +334,167 @@ future<> manager::change_host_filter(host_filter filter) {
+            return make_ready_future<>();
+        } catch (...) {
+            // We cannot map the IP to the corresponding host ID either because
+            // the relevant mapping doesn't exist anymore or an error occurred. Drop it.
```
I reached out to @kostja touching on it and we can go with this approach.

gusev-p replied:
OK

dawmd commented on service/storage_proxy.cc:
```diff
@@ -1019,10 +1033,12 @@ class per_destination_mutation : public mutation_holder {
-    virtual bool store_hint(db::hints::manager& hm, gms::inet_address ep, tracing::trace_state_ptr tr_state) override {
+    virtual bool store_hint(locator::token_metadata_ptr tmptr, db::hints::manager& hm, locator::host_id hid,
+            tracing::trace_state_ptr tr_state) override {
+        const auto ep = tmptr->get_endpoint_for_host_id(hid).value();
```
See https://github.com/scylladb/scylladb/pull/15567#issuecomment-1886719209

gusev-p commented on db/hints/internal/hint_sender.cc:
```diff
@@ -50,8 +50,8 @@ namespace internal {
 
-[[maybe_unused]] std::optional<gms::inet_address> maybe_convert_hid_to_ep_with_erm(
-        locator::effective_replication_map_ptr ermptr, const locator::host_id& hid) {
+std::optional<gms::inet_address> maybe_convert_hid_to_ep_with_erm(locator::effective_replication_map_ptr ermptr,
```
Why we need this helper? It's used only in one place and doesn't bring much value. And in `hint_sender::can_send` we use `get_endpoint_for_host_id_if_known` directly.

gusev-p commented on service/storage_proxy.cc:
```diff
@@ -3193,8 +3206,8 @@ void storage_proxy::register_cdc_operation_result_tracker(const storage_proxy::u
     auto& h = *get_write_response_handler(id);
-
-    size_t hints = hint_to_dead_endpoints(h._mutation_holder, h.get_dead_endpoints(), h._type, h.get_trace_state());
+    const auto ermptr = h._effective_replication_map_ptr;
```
redundant variable `ermptr`

gusev-p commented on db/hints/manager.cc:
```diff
@@ -31,6 +31,7 @@
 #include "replica/database.hh"
 #include "service/storage_proxy.hh"
```
"This commit is part of the series migrating Hinted Handoff from using IPs to using locator::host_id." we usually don't include such preamble, we're already in the PR and reviewers have read its description.

gusev-p commented on db/hints/manager.cc:
```diff
@@ -569,4 +574,87 @@ future<> manager::with_file_update_mutex_for(endpoint_id ep, noncopyable_functio
+            auto maybe_parsed_name = parse_name(de.name);
+
+            if (!maybe_parsed_name) {
+                return remove_directory(current_path, "The directory's name {} is neither IP, nor host ID.");
```
missing `format("...", dir name)`

gusev-p commented on db/hints/resource_manager.cc:
```diff
@@ -146,21 +147,26 @@ void space_watchdog::on_timer() {
-                if (it != shard_manager.ep_managers_end()) {
-                    return with_file_update_mutex(it->second, [this, &shard_manager, dir = std::move(dir), ep_name = std::move(de.name)] () mutable {
-                        return scan_one_ep_dir(dir / ep_name, shard_manager, ep_key_type(ep_name));
+                const locator::host_id hid = locator::host_id{utils::UUID{de.name}};
```
OK

gusev-p commented:
Regarding dtests, I suppose it's our responsibility to make them pass with this PR. We can make a PR to the dtests repository and adjust them so they work with both `host_id`-s and `IP`-s.

avikivity commented:
> Regarding dtests, I suppose it's our responsibility to make them pass with this PR. We can make a PR to the dtests repository and adjust them so they work with both `host_id`-s and `IP`-s.

yes

dawmd commented on db/hints/internal/hint_sender.cc:
```diff
@@ -111,17 +113,24 @@ bool hint_sender::can_send() noexcept {
     }
 
+    const auto tmptr = _shard_manager._proxy.get_token_metadata_ptr();
+    const auto maybe_ep = tmptr->get_endpoint_for_host_id_if_known(_ep_key);
```
A host ID cannot change. Hint senders can only be created towards other nodes, so it's impossible that `_ep_key` points to the local node. I'll add a comment explaining it

dawmd commented on service/storage_proxy.cc:
```diff
@@ -3078,10 +3087,14 @@ storage_proxy::create_write_response_handler_helper(schema_ptr s, const dht::tok
     // If the manager hasn't started yet, no mutation will be performed to another node.
     // No hint will need to be stored.
-    if (_hints_manager.started() && cannot_hint(all, type)) {
+    if (_hints_manager.started() && cannot_hint(all_hids, type)) {
```
I refactored it as part of this PR #16511 since it seems to be a better place for it (or perhaps a completely independent, one-commit PR?)

gusev-p replied:
at your discretion, this doesn't block this PR, of course, just point for improvement

dawmd commented:
Version 3:
* went back to one `on_single_cluster` overload, just with an additional parameter
* no redundant copies of effective replication map
* fix with `format` in `migrate_ip_directories`
* rebased on top of the current master branch

I'll post a PR to dtests in a moment. This PR still needs to be rebased on top of #16511 and hopefully the changes get merged by that time

dawmd commented:
Version 4:
* rebased on top of the current master branch
* ditched the initial commit as it has been covered in #16511

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6142/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7443/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6142/)
#### Failed Tests (6/24107):
- [test_hintedhandoff_removenode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6142/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_removenode) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_removenode)
- [test_hintedhandoff_rebalance](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6142/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_rebalance) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_rebalance)
- [test_hintedhandoff_dont_revive](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6142/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_dont_revive) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_dont_revive)
- [test_hintedhandoff_basic_check](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6142/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_basic_check) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_basic_check)
- [test_tablet_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6142/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_cleanup) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cleanup)
- [topology_experimental_raft.test_tablets.dev.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6142/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_dev_6) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.dev.6)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: spider3.cloudius-systems.com

dawmd commented:
Version 5:
* found a bug in the directory migration function and fixed it (I extended the corresponding dtest to check it as well)
* made logs a little bit more readable (in my opinion)
* some minor modifications in the comments

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6183/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6183/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7485/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6183/)
#### Failed Tests (4/24108):
- [test_hintedhandoff_removenode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6183/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_removenode) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_removenode)
- [test_hintedhandoff_rebalance](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6183/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_rebalance) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_rebalance)
- [test_hintedhandoff_dont_revive](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6183/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_dont_revive) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_dont_revive)
- [test_hintedhandoff_basic_check](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6183/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_basic_check) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_basic_check)
#### Build Details:
- Duration: 1 hr 56 min
- Builder: spider3.cloudius-systems.com

kbr-scylla commented:
Did you investigate the hinted handoff CI test failures?

dawmd commented:
> Did you investigate the hinted handoff CI test failures?

@kbr-scylla they occur because hint directories are now identified by host IDs, while the failed tests rely on their names – them being IPs in particular. The dtests have been adjusted in another PR; ref: https://github.com/scylladb/scylladb/pull/15567#issuecomment-1900448951

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6282/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6282/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6282/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7580/)
#### Failed Tests (4/24163):
- [test_hintedhandoff_removenode](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6282/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_removenode) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_removenode)
- [test_hintedhandoff_rebalance](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6282/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_rebalance) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_rebalance)
- [test_hintedhandoff_dont_revive](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6282/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_dont_revive) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_dont_revive)
- [test_hintedhandoff_basic_check](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6282/testReport/junit/hintedhandoff_additional_test/TestHintedHandoff/Tests___dtest___test_hintedhandoff_basic_check) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_hintedhandoff_basic_check)
#### Build Details:
- Duration: 2 hr 11 min
- Builder: spider3.cloudius-systems.com

kbr-scylla commented:
@dawmd I think you restarted CI too early, before the dtest change got promoted to `master` --- it looks like it's still only on `next` (https://github.com/scylladb/scylla-dtest/commit/e921ae6861313005b1a21f9bdaf7145a650caf77), but CI is running based on `master` IIUC

kbr-scylla commented:
Hm, I don't know why github doesn't show that it's on `master` from that page, but apparently it already is
https://github.com/scylladb/scylla-dtest/commits/master/

I'll restart CI again

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6307/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6307/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6307/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7599/)
#### Build Details:
- Duration: 2 hr 21 min
- Builder: spider6.cloudius-systems.com

gusev-p commented on db/hints/manager.cc:
```diff
@@ -569,4 +576,85 @@ future<> manager::with_file_update_mutex_for(endpoint_id ep, noncopyable_functio
+            
+            if (maybe_parsed_name.value().has_host_id()) {
+                return make_ready_future<>();
+            }
```
I'm not sure how `scan_dir/getdents64` behaves when we mutate the files it's iterating over (renaming/removing). I couldn't find this in the man-s.

I think it would be conceptually simpler to do the migration in phases:
1. readonly run of `scan_dir`, fill `files_to_rename/files_to_remove` vectors;
2. run all renames, append all the failures to the `files_to_remove`;
3. run all removes;
4. run one single `co_await io_check(sync_directory, _hints_dir.native());` when we are done.

avikivity replied:
Sensible. It's also something that is liable to depend on the filesystem and can change with implementations.

gusev-p commented on db/hints/manager.cc:
```diff
@@ -569,4 +576,85 @@ future<> manager::with_file_update_mutex_for(endpoint_id ep, noncopyable_functio
+            // or leave the directory. Simply remove it.
+            return rename_file(current_path.native(), hid_path.native()).then([current_path, hid_path] {
+                manager_logger.info("Successfully renamed directory {} to {}", current_path.filename(), hid_path.filename());
+            }).handle_exception([remove_directory, current_path] (auto&& e) {
```
It's better to use coroutines in new code, they are more readable than `then/handle_exception/etc`

dawmd replied:
We can't use `co_await` in `catch` clauses, but I'll see if we can use more C++20 code after changing the migration phases

bhalevy commented on db/hints/manager.cc:
```diff
@@ -352,10 +371,15 @@ bool manager::store_hint(endpoint_id ep, schema_ptr s, lw_shared_ptr<const froze
     // There is no need to check the DC here because if there is an in-flight hint for this
     // endpoint, then this means that its DC has already been checked and found to be ok.
-    return _stats.size_of_hints_in_progress > MAX_SIZE_OF_HINTS_IN_PROGRESS
+    if (!(_stats.size_of_hints_in_progress > MAX_SIZE_OF_HINTS_IN_PROGRESS
```
Yes, that was accidental.
Feel free to write a test for this function.

piodul replied:
Just to clarify: the PR inverts the condition because now the `if` enters a branch which returns `false`. Benny's commit didn't invert the whole `if`, only one of the conjuncts, i.e. there is a missing `!` before `_proxy.local_db().get_token_metadata().get_topology().is_me(ep)`.

I created an issue: https://github.com/scylladb/scylladb/issues/17636

As for writing a test... with error injection blocking storing hints to disk it should be possible to write a test indeed.

gusev-p replied:
I also created an issue but forgot to post it here https://github.com/scylladb/scylladb/issues/17129. Will close as duplicate

avikivity commented on service/storage_proxy.cc:
```diff
@@ -123,6 +123,19 @@ seastar::metrics::label_instance current_scheduling_group_label() {
 
 }
 
+[[maybe_unused]] static locator::host_id convert_ep_to_hid_with_erm(locator::effective_replication_map_ptr ermptr,
```
This name has the highest density of acronyms/contractions in the system. How is a newcomer supposed to understand it?

avikivity commented on service/endpoint_lifecycle_subscriber.hh:
```diff
@@ -66,7 +67,7 @@ public:
 
     future<> notify_down(gms::inet_address endpoint);
-    future<> notify_left(gms::inet_address endpoint);
+    future<> notify_left(gms::inet_address endpoint, locator::host_id hid);
```
It would be better to convert all the callbacks to use host ID exclusively, but I understand it's not practical. It's okay to have this redundant call convention, but let's plan to convert everything later.

avikivity commented on db/hints/internal/hint_sender.cc:
```diff
@@ -80,9 +80,20 @@ future<> hint_sender::do_send_one_mutation(frozen_mutation_and_schema m, locator
-        if (boost::range::find(natural_endpoints, end_point_key()) != natural_endpoints.end()) {
+        const auto maybe_addr = [&] () -> std::optional<gms::inet_address> {
+            const auto& tm = ermp->get_token_metadata();
+            if (auto maybe_ep = tm.get_endpoint_for_host_id_if_known(end_point_key())) {
```
tangent observations:
  - since we look up the endpoint for every mutation, this is already safe for tablet migration while the node was down
  - but we need to extend the hold period of effective_replication_map_ptr until we get the response; it acts as a lock that prevents a tablet from migrating while the mutation is in-flight
 
Please follow up with extending the lifetime. Note that what we have now still works - @gusev-p will reject the mutation if it arrives with outdated topology. But it's better not to have to reject it.

avikivity replied:
Another tangent: some coroutinizations would be helpful around here.

avikivity commented on db/hints/internal/hint_sender.cc:
```diff
@@ -80,9 +80,20 @@ future<> hint_sender::do_send_one_mutation(frozen_mutation_and_schema m, locator
+                return tm.get_topology().my_address();
+            }
+            return {};
+        } ();
```
Style note: I prefer std::invoke(long lambda) to (long lambda)(). The former syntax immediately tells me I'm looking at a lambda introduced for a scope, while the latter syntax keeps me thinking we're defining a helper until we see the invocation (or don't see it since it's just three characters).

avikivity commented on db/hints/manager.hh:
```diff
@@ -317,6 +317,15 @@ private:
+    /// this purpose. It's only necessary when upgrading Scylla.
+    ///
+    /// This function should ONLY be called by `manager::start()`.
+    future<> migrate_ip_directories();
```
What about failed upgrades?

Scenario:
1. upgrade a node
2. node migrates its hint directories
3. cold feet, downgrade
4. old version confused

Better to defer until the entire cluster is upgraded using a feature. This means we need to be able to replay using either format. The alternative is to give up on this, and just make sure older versions know to ignore the new format.

dawmd replied:
@mykaul a node could be confused because it comes across a directory whose name doesn't represent a valid IP address, in which case there will be problems in the current code. That situation can happen when some of the nodes in the cluster get upgraded and start writing hints on disk, naming the directories using host IDs, but at some point the upgrade of another node fails. If I understand the documentation correctly, the whole cluster will undergo downgrade in that scenario. The created hint directories will remain, though.

I also think that out of those two approaches, ignoring those directories would be better. It should require far fewer changes and perhaps prevent introducing new bugs (compared to the former idea). However, if we cannot assume that the version the cluster is being upgraded from is the latest minor of the previous major, then (if I'm not mistaken) we can't simply add code responsible for ignoring those directories. If that's the case, a feature flag is the only solution I can see at the moment (even specifying in the documentation that hinted handoff should be disabled before the whole cluster has been upgraded wouldn't help because the hint managers for MVs will be created anyway).

However, if we _can_ add new code to the previous versions and assume the user uses it, then I think the easiest solution would be to do the exact same thing as in this PR – we try to map directories from host IDs to IPs if Hinted Handoff is still based on IPs, or the other way round if it's based on host IDs. That would boil down to cherry picking one commit (perhaps modified a bit if parameterising the code by template-like stuff makes it too complex).

bhalevy replied:
Requiring a minor upgrade before the major upgrade is a no-go IMO.

The new version is required to support ip-address based view hints, right?
So it needs this code path in any case and a cluster feature to enable host_id-based hint directories makes the most sense.

mykaul replied:
I'm somewhat surprised how current code doesn't ignore directories whose name is not a valid IP address.

kostja replied:
Please let's not call this pull request to death because of the potential upgrade failure. If the ugprade fails, we lose hints. What is unacceptable with this? Please keep in mind hints are off by default now.

kostja replied:
@avikivity

dawmd replied:
Possible solutions:
#### 1. Use a feature flag ####
Continue storing hints naming directories after IPs and only after the whole cluster has been upgraded, switch to naming them after host IDs. The biggest advantage of this approach is that no hints should be lost in the process. However, a price comes with it: between the start and end of an upgrade, we will have to store hints in directories named after IPs. When the upgrade finishes, we should rename them to host IDs. This part is tricky because new hints might still be generated, so hinted handoff would have to be turned off for some time to be able to move those directories. Also, it would make the code more complex because we'd have to handle both IPs and host IDs; and as of now, I can't give an accurate estimate of how much would have to be modified.

#### 2. Stop storing hints between the start and the end of an upgrade ####
Alternatively, when upgrading a node, we could disable hinted handoff when an upgrade starts and only when the whole cluster finishes the upgrade would we rename the directories and enable it back. This way all of the logic would correspond to using host IDs and the changes should be minimal. However, since hints would not be stored during an upgrade and that applies to MVs as well of course.

I discussed it with my team and we opt for the latter approach.

kostja replied:
@avikivity this comment has been left unaddressed for two weeks because it's not clear how to address it. 

In my view it doesn't need addressing in this series at all. This series started when the upgrade path to 6.0 was non-existent. So the upgrade problem for topology was solved while this series was being developed. Therefore I believe this series can follow hte same pattern and provide an upgrade path in a follow up.

But most importantly, I don't see the issue. Cold feet? Downgrade? Old version doesn't start because of the unknown files? Delete the hint files & restart. 

Keep in mind that hints are off by default. The chance of stepping into the scenario you describe is very low.

kostja replied:
@avikivity all in all with @dawmd suggestions and the suggestion to backpatch 5.x to ignore unknown files (it probably already does, so perhaps no back patching is needed), we have 4 options to choose from.

Please be so kind and choose a solution so that this pull request can get unblocked.

avikivity replied:
Hints are required for materialized views to work correctly, and hints are written exactly during upgrades.

I think this merits a cluster feature.

bhalevy commented:
needs rebasing

dawmd commented:
Version 6:
* I try to avoid too many acronyms and abbreviations when naming functions and variables
* I briefly explain the process of migration in the documentation
* renaming of hint directories is performed more safely, as per Petr's suggestion. The code becomes cleaner and is easier to understand. It's also coroutinized
* Scylla creates hint directories representing IP addresses by default. The migration only happens when the corresponding cluster feature is enabled
* smaller changes in resource manager, e.g. coroutinization

Points for possible improvement:
* [24bbfd2](https://github.com/scylladb/scylladb/pull/15567/commits/24bbfd21c65b45af3ab1a6efe45bbe27c6537616) introduces a data structure keeping track of mappings between host IDs identifying endpoint managers and the IPs that their managed hint directories represent.
It's a hacky solution, but out of the options I've considered, it seems the cleanest and simplest way to solve an issue with `resource_manager`: when a hint directory stores more than two hints at once, we want to mark it as a directory with pending hints. Depending on how it's being handled, some endpoint managers might be stopped for example.
If we assume hint directories still represent IP addresses, problems may arise. Endpoint managers are identified by host IDs now and since `locator::token_metadata` can change over time, it may happen that trying to map an IP using it will lead to blocking the wrong endpoint manager – if it's managing a different directory.
Because of that, we need a means to associate a host ID that an endpoint manager is identified by with the IP address of the directory it manages. That's exactly why the data structure is used.

I'll try to update the upgrade dtest as soon as possible to adjust it to the current semantics of the migration.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7668/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7668/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/661/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8689/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7668/)
#### Build Details:
- Duration: 2 hr 25 min
- Builder: spider7.cloudius-systems.com

dawmd commented:
Version 7:
* fixed minor bugs:
  * we set `_uses_host_id` to true as soon as the migration process has finished. We previously didn't do it
  * we don't allow for `drain_for` or creating a sync point to proceed while the migration process is running.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7788/)
#### Build Details:
- Duration: 1 hr 10 min
- Builder: i-0e1d9236bc4fd437e (m5d.8xlarge)

dawmd commented on db/hints/internal/hint_sender.cc:
```diff
@@ -95,17 +103,27 @@ bool hint_sender::can_send() noexcept {
+    // A host ID cannot change. Hint senders can only be created towards other nodes,
+    // so it's impossible that _ep_key points to the local node.
+    // That's why we don't need to check `tmptr->get_topology().is_me(_ep_key)`.
+    const auto maybe_ep = tmptr->get_endpoint_for_host_id_if_known(_ep_key);
```
Good point, I'll do it. I'll modify the signature of the function in a separate PR

dawmd commented:
Version 8:
* `resource_manager` doesn't keep track of its state – migrating or non-migrating. We expose its semaphore and access it directly from the shard hint manager instead,
* we don't unnecessarily allocate memory in `can_hint_for()` and `too_many_in_flight_hints_for`
* the documentation has been modified to highlight Scylla has already started using host-ID-based hinted handoff,
* indentation fixes have been moved to separate commits to not bring chaos,
* calls to `hint_endpoint_manager::stop()` in `perform_migration()` might've thrown an exception if a draining process had already been started; we introduce more state in the hint shard manager to keep track of those endpoint managers and await them separately.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7944/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7944/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/909/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/8937/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/7944/)
#### Build Details:
- Duration: 2 hr 22 min
- Builder: spider8.cloudius-systems.com

kbr-scylla commented:
@dawmd please select a subset of 3 reviewers, including at least 1 maintainer, and deselect everyone else in the Reviewers list.

piodul commented on db/hints/manager.cc:
```diff
@@ -450,25 +449,48 @@ future<> manager::change_host_filter(host_filter filter) {
-            if (_ep_managers.contains(ep) || !_host_filter.can_hint_for(topology, ep)) {
-                return make_ready_future();
-            }
+    /* RAII lock for token metadata */ {
```
Is this needed? The token metadata lock is already limited by the scope of `try { ... }`.

dawmd commented on service/storage_proxy.cc:
```diff
@@ -123,13 +123,17 @@ seastar::metrics::label_instance current_scheduling_group_label() {
+    // During initialization of Scylla, there is a time when locator::token_metadata doesn't have any mappings yet
+    // or it is unstable (it removes and adds mappings IP<->host ID multiple times). However, this function provides
+    // a safe way to check if a given IP or host ID belongs to this node, even during that time. Hence this additional
+    // check.
```
That use case was necessary in the previous version of the code because in `storage_proxy::create_write_response_handler()` we were explicitly creating a new vector of host IDs, obtaining them by mapping the IPs we got from the effective replication map. Then the vector was being passed to `cannot_hint()`. We have to perform queries to system tables during early stages of start-up, when topology is still unstable; because each mutation has to go through this function, if trying to map the IP of the local node failed (as it could!), Scylla would crash.

However, in the current version of this PR (where we do NOT create the vector anymore. We only create a view), and thanks to #16511 having been merged, I think the function is completely redundant. I don't know the semantics of Boost adaptors in detail, but I can modify the code so that it uses `std::ranges` to guarantee lazy evaluation of views. The function `cannot_hint()` first verifies that hinted handoff has been started (cf. the mentioned issue), so until that happens, we wouldn't need to perform any mapping at all, and there would be no problem.

dawmd commented on db/hints/manager.cc:
```diff
@@ -414,18 +416,55 @@ bool manager::store_hint(endpoint_id host_id, gms::inet_address ip, schema_ptr s
+static bool endpoint_downtime_not_bigger_than(const gms::gossiper& gossiper, const locator::host_id& host_id,
+        uint64_t max_downtime_us)
+{
+    // We want to enforce small buffer optimization in the call
+    // to `gms::gossiper::for_each_endpoint_state_until()` below
+    // to avoid an unnecessary allocation.
+    // Since we need all these four pieces of information in the lambda,
+    // the function object passed to the function might be too big.
+    // That's why we create it locally on the stack and only pass a reference to it.
+    struct sbo_info {
+        locator::host_id host_id;
+        const gms::gossiper& gossiper;
+        int64_t max_hint_window_us;
+        bool small_node_downtime;
+    };
+
+    sbo_info info {
+        .host_id = host_id,
+        .gossiper = gossiper,
+        .max_hint_window_us = max_downtime_us,
+        .small_node_downtime = false
+    };
+
+    gossiper.for_each_endpoint_state_until(
+            [&info] (const gms::inet_address& ip, const gms::endpoint_state& state) {
+        const auto app_state = state.get_application_state_ptr(gms::application_state::HOST_ID);
+        const auto host_id = locator::host_id{utils::UUID{app_state->value()}};
+        if (!app_state || host_id != info.host_id) {
+            return stop_iteration::no;
+        }
+        if (info.gossiper.get_endpoint_downtime(ip) <= info.max_hint_window_us) {
+            info.small_node_downtime = true;
+            return stop_iteration::yes;
+        }
+        return stop_iteration::no;
+    });
```
The lambda is just one address. I don't think compilers are enforced to apply SBO here, but I'm pretty sure both Clang and GCC would. There's no advantage of not doing that. But if you think it's still unnecessary code bloat, I can absolutely understand that and I'll drop the commit

dawmd commented on db/hints/manager.cc:
```diff
@@ -768,4 +801,62 @@ future<> manager::migrate_ip_directories() {
+    //      boil down to not initializing some of the endpoint managers. No data corruption
+    //      is possible.
+    if (_state.contains(state::stopping) || _state.contains(state::draining_all)) {
+        co_return;
```
It's not. If we're stopping the node, then we drop the idea of migrating hinted handoff – there's neither time nor need for that. The other case is less clear because there is no clear semantics of draining the local node – it's not specified when exactly it happens. However, in practice it only occurs when the node is leaving the cluster, which boils down to the same thing – stopping it.
I can add a comment explaining this in the code.

dawmd commented on db/hints/manager.cc:
```diff
@@ -768,4 +801,62 @@ future<> manager::migrate_ip_directories() {
+        }
+    }
+    _drained_endpoint_managers.clear();
+    _ep_managers.clear();
```
Yes, thank you

dawmd commented on db/hints/manager.cc:
```diff
@@ -267,6 +281,16 @@ sync_point::shard_rps manager::calculate_current_sync_point(std::span<const gms:
 }
 
 future<> manager::wait_for_sync_point(abort_source& as, const sync_point::shard_rps& rps) {
+    // During the migration process, endpoint managers are all stopped. They're resumed
+    // by the end of it, so in case we call this function while hinted handoff is being
+    // migrated to host IDs, we need to wait for it to end to avoid race conditions.
+    //
+    // This scenario is quite unlikely because when stopping an endpoint manager,
+    // the sync points should all be invalidated and exceptions should be issued,
+    // but it's better to be safe. The danger mostly lies in the part where we're iterating
+    // over endpoint managers.
+    co_await _migrating_done.get_future();
```
You're right. But if `coroutine::parallel_for` is not preemptive, then shouldn't this work? Thanks to this await, it's guaranteed that when we proceed to the next instruction in the function, there is no ongoing migration. Once we start the lambda for all endpoint managers, we're safe. If a migration starts afterwards, but before the coroutine finishes, all hint senders will be stopped and `wait_until_hints_are_replayed_up_to()` will return an exception indicating an abort.

piodul replied:
Yes, I also think that it will work _right now_, but relying on such non-local assumptions makes the code fragile. Somebody might change that code later and the code here will break. Relying on explicit synchronization primitives is better because they give a stronger indication to the people that they should think twice before removing them - people don't give as much attention to (non)preemptiveness.

dawmd commented on db/hints/manager.cc:
```diff
@@ -326,6 +329,21 @@ future<> manager::wait_for_sync_point(abort_source& as, const sync_point::shard_
+        // If there is no mapping in `_hint_directory_manager` corresponding to either `host_id`, or `ip`,
+        // we need to create a new endpoint manager.
+        _hint_directory_manager.insert_mapping(host_id, ip);
+    }
```
No, it works like that by design.

The main problem we need to solve is associate endpoint managers (identified by host IDs) with hint directories (identified by IPs before an upgrade). If we were to try to keep a 1-1 correspondence between the two, we'd need to keep renaming hint directories every time a node changes its IP or a replace operation is performed. We could do it, but we'd incur additional complexity (mostly due to the activity of resource manager and space watchdog).

That's why I decided to associate possibly multiple target nodes with the same hint directory. A simple scenario is like this:
* node A has IP I
* A changes its IP to I'
* node B joins the cluster with IP I
Then, we will save hints towards both node A and node B in the same hint directory because of the mapping kept in `_hint_directory_manager`. Replacing a node is a similar situation except for the fact we get rid of one of the nodes. And here comes the bad side of the story.

If it happens that we associate two different nodes – node A and node B – with the same hint directory and one of those nodes – e.g. node A – leaves the cluster, we will start draining the corresponding endpoint manager – meaning, incoming hints towards node B will be rejected until the draining process finishes, even though we should save those hints. When it's done, we will simply create a new endpoint manager in place of the previous one.

Another problem is we might overly limit the concurrency if we save hints towards multiple nodes via the same endpoint manager.

However, the most important thing is to make sure no data races will occur. And they won't because of the invariant we keep, i.e. at any point in time, there is always exactly one endpoint manager corresponding to a given hint directory.

It's not the smartest way to go about it, but I think it's a low price for making things work until the upgrade process is finished. Once it's done, everything will go back to how it was before this patch, i.e. we regain the 1-to-1 correspondence between endpoint managers and hint directories.

dawmd commented:
Version 9:
* Added more comments in places that might not be obvious.
* Elaborated on the changes in `change_host_filter()` in the commit message
* Scylla logs the reason why renaming a directory has failed.
* We rename and remove hint directories in parallel.
* Dropped `_drained_endpoint_managers`. We start using the drain lock instead.
* Turned the auxiliary function in storage proxy to a lambda. It was only called in one place and it's easier to understand its purpose now.

I explained my reasoning why the current use of `_hint_directory_manager` is safe in the comments above. I forgot to apply https://github.com/scylladb/scylladb/pull/15567#discussion_r1567137023. I will go back to it on Friday.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:x: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8262/)
#### Build Details:
- Duration: 14 min
- Builder: spider4.cloudius-systems.com

dawmd commented:
Version 10:
* Rebased on top of the current master (necessary because of the changes in logging).
* Replaced the logic in `manager::wait_for_sync_point`: we don't rely on details of the current implementation. Instead, we introduce a shared mutex for that purpose and use it explicitly.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8359/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8359/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9270/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1249/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/8359/)
#### Build Details:
- Duration: 2 hr 32 min
- Builder: spider2.cloudius-systems.com

piodul commented on db/hints/manager.cc:
```diff
@@ -269,6 +284,19 @@ sync_point::shard_rps manager::calculate_current_sync_point(std::span<const gms:
+    using lock_type = std::shared_lock<seastar::shared_mutex>;
+    // Prevent the migration to host-ID-based hinted handoff until this function finishes its execution.
+    const auto shared_lock = co_await std::invoke(coroutine::lambda([&] () -> future<lock_type> {
+        co_await _migration_mutex.lock_shared();
```
Nit: I think it would be a bit nicer to use `seastar::with_shared` here instead.

Not worth fixing it now.

piodul commented:
Queued.

@dawmd next time please use full paths to issues/prs in the PR description and commit descriptions (e.g. `scylladb/scylladb#1234` instead of just `#1234`). It helps avoid confusion in scylla-enterprise. This time, I fixed the ones in the description when I was merging.
